libraries/chain/block_log.cpp

```diff
@@ -388,27 +451,44 @@ namespace eosio { namespace chain {
       block_file << chain_id;
    }
 
-   signed_block_ptr block_log::read_block(uint64_t pos)const {
-      my->check_open_files();
+   pruned_block_ptr detail::block_log_impl::read_log_entry() {
+      auto ds = block_file.create_datastream();
+      if (version >= 4) {
+         pruned_block_ptr result = std::make_shared<pruned_block>();
+         detail::block_log_impl::unpack(ds, *result);
+         return result;
+      } else {
+         signed_block_ptr result = std::make_shared<signed_block>();
+         fc::raw::unpack(ds, *result);
+         return std::make_shared<pruned_block>(*result, true);
+      }
+   }
 
+   pruned_block_ptr block_log::read_block(uint64_t pos) const {
+      my->check_open_files();
       my->block_file.seek(pos);
-      signed_block_ptr result = std::make_shared<signed_block>();
-      auto ds = my->block_file.create_datastream();
-      fc::raw::unpack(ds, *result);
-      return result;
+      return my->read_log_entry();
    }
 
-   void block_log::read_block_header(block_header& bh, uint64_t pos)const {
+   void block_log::read_block_header(block_header& bh, uint64_t pos) const {
       my->check_open_files();
 
+      if (my->version == 4)
+         pos += 2 * sizeof(uint32_t);
```

Comment:
Instead of just skipping over the appropriate amount, actually unpack the length and the compression status. Then assert that the compression status is `compression_type::none`, because this implementation doesn't yet support compressed block log entries yet.